src/api/oauth2.rs

```diff
@@ -0,0 +1,293 @@
+    let code = get_code(&("http://localhost".to_string() + redirect_url));
+
+    let message = "Go back to your terminal :)";
+    let response = format!(
+        "HTTP/1.1 200 OK\r\ncontent-length: {}\r\n\r\n{}",
+        message.len(),
+        message
+    );
+    stream
+        .write_all(response.as_bytes())
+        .map_err(|_| OAuthError::AuthCodeListenerWrite)?;
+
+    code
+}
+
+// If the specified `redirect_uri` is HTTP, loopback, and contains a port,
+// then the corresponding socket address is returned.
+fn get_socket_address(redirect_uri: &str) -> Option<SocketAddr> {
+    let url = match Url::parse(redirect_uri) {
+        Ok(u) if u.scheme() == "http" && u.port().is_some() => u,
+        _ => return None,
+    };
+    let socket_addr = match url.socket_addrs(|| None) {
+        Ok(mut addrs) => addrs.pop(),
+        _ => None,
+    };
+    if let Some(s) = socket_addr {
+        if s.ip().is_loopback() {
+            return socket_addr;
+        }
+    }
+    None
+}
+
+/// Obtain a Spotify access token using the authorization code with PKCE OAuth flow.
+/// The redirect_uri must match what is registered to the client ID.
+pub fn get_access_token(
+    client_id: &str,
+    redirect_uri: &str,
+    scopes: Vec<&str>,
+) -> Result<OAuthToken, OAuthError> {
+    let auth_url = AuthUrl::new("https://accounts.spotify.com/authorize".to_string())
+        .map_err(|_| OAuthError::InvalidSpotifyUri)?;
+    let token_url = TokenUrl::new("https://accounts.spotify.com/api/token".to_string())
+        .map_err(|_| OAuthError::InvalidSpotifyUri)?;
+    let redirect_url =
+        RedirectUrl::new(redirect_uri.to_string()).map_err(|e| OAuthError::InvalidRedirectUri {
+            uri: redirect_uri.to_string(),
+            e,
+        })?;
+    let client = BasicClient::new(
+        ClientId::new(client_id.to_string()),
+        None,
+        auth_url,
+        Some(token_url),
+    )
+    .set_redirect_uri(redirect_url);
+
+    let (pkce_challenge, pkce_verifier) = PkceCodeChallenge::new_random_sha256();
+
+    // Generate the full authorization URL.
+    // Some of these scopes are unavailable for custom client IDs. Which?
+    let request_scopes: Vec<oauth2::Scope> = scopes
+        .clone()
+        .into_iter()
+        .map(|s| Scope::new(s.into()))
+        .collect();
+    let (auth_url, _) = client
+        .authorize_url(CsrfToken::new_random)
+        .add_scopes(request_scopes)
+        .set_pkce_challenge(pkce_challenge)
+        .url();
+
+    println!("Browse to: {}", auth_url);
+    if let Err(err) = open::that(auth_url.to_string()) {
+        eprintln!("An error occurred when opening '{}': {}", auth_url, err)
+    }
+
+    let code = match get_socket_address(redirect_uri) {
+        Some(addr) => get_authcode_listener(addr),
+        _ => get_authcode_stdin(),
+    }?;
+    trace!("Exchange {code:?} for access token");
+
+    // Do this sync in another thread because I am too stupid to make the async version work.
```

Comment:
that's absolutely fine that can be adressed later :smile:

Reply:
:laughing: https://github.com/librespot-org/librespot/blob/82076e882f3cdebec863a3c0aa79888ec47b3c76/oauth/src/lib.rs#L212

as much as I'd love to use that file directly, it's just `println`'ing the URL and IDK how to consume that from a library that's writing to the consuming applications std-out

Reply:
I've made some changes in regards to the CSRF token. I will try and get  PR in to librespot for this as it makes more sense to me at least to try and use the default browser...